server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,164 @@
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.bulk.BulkRequestBuilder;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.synonyms.PutSynonymsAction;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.DeleteByQueryAction;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.indices.SystemIndexDescriptor;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+import static org.elasticsearch.index.mapper.MapperService.SINGLE_MAPPING_NAME;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+
+public class SynonymsManagementAPIService {
+    public static final String SYNONYMS_INDEX = ".synonyms";
+    public static final String SYNONYMS_ORIGIN = "synonyms";
+
+    public static final String SYNONYMS_FEATURE_NAME = "synonyms";
+    public static final String SYNONYM_SET_FIELD = "synonym_set";
+    public static final String SYNONYM_FIELD = "synonym";
+
+    private final Client client;
+
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX + "*")
+        .setDescription("Synonyms index for synonyms managed through APIs")
+        .setPrimaryIndex(SYNONYMS_INDEX)
+        .setMappings(mappings())
+        .setSettings(settings())
+        .setVersionMetaKey("version")
+        .setOrigin(SYNONYMS_ORIGIN)
+        .build();
+
+    public SynonymsManagementAPIService(Client client) {
+        // TODO Should we set an OriginSettingClient? We would need to check the origin at AuthorizationUtils if we set an
+        this.client = new OriginSettingClient(client, SYNONYMS_ORIGIN);
+    }
+
+    private static XContentBuilder mappings() {
+        try {
+            XContentBuilder builder = jsonBuilder();
+            builder.startObject();
+            {
+                builder.startObject(SINGLE_MAPPING_NAME);
+                {
+                    builder.startObject("_meta");
+                    {
+                        builder.field("version", Version.CURRENT.toString());
+                    }
+                    builder.endObject();
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject("synonym");
+                        {
+                            builder.field("type", "text");
```

Comment:
may be use `match_only_text`, as I don't think we will ever need scoring.

Reply:
I'm ok with that. Will using `match_only_text` prevent us from using "prefix", "wildcard", or "regexp" queries though? That might be useful for providing a search experience...

We might as well choose not to pre-optimize on this and decide later...

Reply:
No, it doesn't; you can do "prefix", "wildcard", or "regexp" queries with `match_only_text` as well.

Reply:
@carlosdelest What is the decision on this comment, should we use `text` or `match_only_text` should be enough?

Reply:
I've changed it to `match_only_text`. Thanks for the discussion!